### ISSUE TITLE
Arbitrary objective functions and noise models

### ISSUE THREAD

**Issue by paulstapor** [AMICI-dev/AMICI#1259]
Currently, the objective function (internally `Jy`) has the following derivative implemented:
`D(Jy)/Dp = dJydy * (dydp + dydx * sx) + dJydsigmay * dsigmaydp`

I was wondering,  what's the advantage of having this intermediate step `sigmay`. What would go wrong (would anything go wrong?) or would it have any drawbacks to have a general `dJydp` there? In principle, this would allow more general objective functions, right?

Just wanted to discuss, whether it might make sense to change this at some point.

**Comment by yannikschaelte** [comment]
Sounds reasonable. Would there be advantages of having dsigmaydp + dJydsigmay?

**Comment by lcontento** [comment]
> Would there be advantages of having dsigmaydp + dJydsigmay?

I don't see any (even if there are sigma expressions used multiple times the compiler should be smart enough to optimize them away). However, in the case of noise parameters which are set on a by-measurement basis it may be helpful to keep sigmas separated (maybe the code that sets the noise parameter values cannot be easily expressed as a sympy expression, maybe substituting it directly into the objective would lead to bloated expressions that slow down model import).

**Comment by dweindl** [comment]
Yes, what Lorenzo said. With having not only `dJydp`, it's currently pretty easy to just skip the entries for the constant sigmas. Not sure how a good alternative implementation would look if we had only `dJydp`.

**Comment by paulstapor** [comment]
Okay, then it might be reasonable to keep the current implementation and add a term for dJydp  which excludes sigmas. Should be pretty easy to implement...
The question is, if this is necessary at all. Not sure, whether we have cases, which are currently not covered. But it's not extremely difficult to make such cases up... Might be a pitfall for some users...
